### PR TITLE
config: Clean up hwloc options

### PR DESCRIPTION
**Pull request by raffenet**
Make hwloc configure options in MPICH more consistent with established
options in Hydra. Use PAC_CHECK_PREFIX for setting appropriate flags
and delete extra cruft. Fixes linking MPICH with an external hwloc
installation.

### CODE REVIEW

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by Komdosh** [comment]
When this will be merged?

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
@hzhou this probably has some overlap with #3430, but might be worth looking at in the short term to fix some existing issues.

File: configure.ac [hunk 1]
```diff
@@ -1254,16 +1254,8 @@ fi
...
-AC_ARG_WITH([hwloc-prefix],
-            [AS_HELP_STRING([[--with-hwloc-prefix[=DIR]]],
-                            [use the HWLOC library installed in DIR,
-                             rather than the one included in src/hwloc.  Pass
-                             "embedded" to force usage of the HWLOC source
-                             distributed with MPICH.])],
-            [],dnl action-if-given
-            [with_hwloc_prefix=embedded]) dnl action-if-not-given
+# or system path)
+PAC_CHECK_PREFIX(hwloc)
```
Comment (hzhou) [inline configure.ac]: `PAC_CHECK_PREFIX` has the same functionality as `PAC_SET_HEADER_LIB_PATH` except the former adds `-prefix` to the option name, right? We used the latter for `libfabric` and `ucx`. I think at least we should be consistent and use a single style. 

reference: https://github.com/pmodels/mpich/pull/4118/files

Comment (raffenet) [reply configure.ac]: True. I was at least trying to be consistent with Hydra (https://github.com/pmodels/mpich/blob/c29531ab34c1d0e8029f2a834a52a92d67024862/src/pm/hydra/configure.ac#L421) :smile:, which had hwloc integration before MPICH. I think we can safely change both of them, though.

**Comment by Komdosh** [comment]
It is all set, could you merge that?

**Comment by raffenet** [comment]
@Komdosh there was a request to modify this PR to align the hwloc options with other libraries in MPICH configure. An update will be coming sometime soon.

**Comment by eschnett** [comment]
This patch is necessary for me, since I'm using mpich with hwloc, and when I use the internal hwloc its symbols are still exported, leading to conflict.

File: configure.ac [hunk 2]
```diff
@@ -1291,30 +1281,17 @@ elif test "$with_hwloc_prefix" = "embedded" ; then
...
       have_hwloc=yes
+      PAC_PREPEND_FLAG([-lhwloc], [WRAPPER_LIBS])
    else
       have_hwloc=no
    fi
    AC_MSG_RESULT([$have_hwloc])
-
-   # FIXME: Disable hwloc on Cygwin for now. The hwloc package,
-   # atleast as of 1.0.2, does not install correctly on Cygwin
-   AS_CASE([$host], [*-*-cygwin], [have_hwloc=no])
```
Comment (pavanbalaji) [inline configure.ac]: This part is unrelated to this patch.  If we really want to not disable hwloc on cygwin, that should be a separate patch and tested.

**Comment by hzhou** [comment]
> This patch is necessary for me, since I'm using mpich with hwloc, and when I use the internal hwloc its symbols are still exported, leading to conflict.

I am surprised the embedded hwloc symbols are visible to the outside. Could you verbosely describe your build process and maybe a simple code that uses external hwloc that exposes the issue?

**Comment by eschnett** [comment]
@hzhou I am building mpich via Spack https://spack.io. Spack keeps a log when it build a package, so I can give you some information.

I am building mpich 3.3.2 on OS X Catalina using gcc 9.2.0.

Configure arguments:
```
==> [2020-01-26-15:24:23.042045] '/var/folders/gl/zvl8d6415vsbkd50nnll95k40000gs/T/eschnett/spack-stage/spack-stage-mpich-3.3.2-kagnmitv33crplu2epdnwfqrcwsobkuf/spack-src/configure' '--prefix=/Users/eschnett/src/spack-develop/opt/spack/darwin-catalina-skylake/gcc-9.2.0-spack/mpich-3.3.2-kagnmitv33crplu2epdnwfqrcwsobkuf' '--enable-shared' '--with-pm=hydra' '--enable-romio' '--without-ibverbs' '--enable-wrapper-rpath=yes' '--with-slurm=no' '--with-pmi=simple' '--with-device=ch3:nemesis:tcp'
```

The configure output contains these hwloc-related lines (among many others):
```
checking hwloc building mode... embedded
configure: hwloc builddir: /private/var/folders/gl/zvl8d6415vsbkd50nnll95k40000gs/T/eschnett/spack-stage/spack-stage-mpich-3.3.2-kagnmitv33crplu2epdnwfqrcwsobkuf/spack-src/src/hwloc
configure: hwloc srcdir: /var/folders/gl/zvl8d6415vsbkd50nnll95k40000gs/T/eschnett/spack-stage/spack-stage-mpich-3.3.2-kagnmitv33crplu2epdnwfqrcwsobkuf/spack-src/src/hwloc
configure: Detected VPATH build
checking for hwloc version... 2.0.3rc2-git
checking if want hwloc maintainer support... disabled (embedded mode)
checking for hwloc directory prefix... src/hwloc/
checking for hwloc symbol prefix... hwloc_
```

The generated static library `libpmpi.a` contains many hwloc symbols:
```
$ nm libpmpi.a | grep hwloc | tail -n 10
00000000000002f0 T _hwloc_obj_type_is_icache
00000000000002c0 T _hwloc_obj_type_is_io
00000000000002b0 T _hwloc_obj_type_is_memory
00000000000002a0 T _hwloc_obj_type_is_normal
0000000000000b90 T _hwloc_obj_type_snprintf
00000000000004b0 T _hwloc_obj_type_string
                 U _hwloc_pci_class_string
                 U _hwloc_snprintf
0000000000000630 T _hwloc_type_sscanf
0000000000000ae0 T _hwloc_type_sscanf_as_depth
```

I link my application against both mpich and against hwloc 2.1.0. Since I've been burned by inconsistent build-time/run-time versions before, I added this function to my application to check:
```
int hwloc_version(void) {
  const char *library_version =
  // HWLOC_VERSION appears first in hwloc 2.1, before that one has to consult
  // the output of hwloc-info --version or pkgconfig
#if defined(HWLOC_VERSION) // modern hwloc
  HWLOC_VERSION;
#elif defined(HWLOC_VERSION_BUILD) // bundled copy
  HWLOC_VERSION_BUILD;
#elif defined(HWLOC_VERSION_PKGCONFIG)  // found via pkgconfig
  HWLOC_VERSION_PKGCONFIG;
#elif defined(HWLOC_VERSION_HWLOCINFO) // found hwloc-info
  HWLOC_VERSION_HWLOCINFO;
#else
  "unknown";
#endif
  unsigned buildtime_api_version = HWLOC_API_VERSION;
  unsigned runtime_api_version = hwloc_get_api_version();
  // TODO: Check only major version number?
  if (runtime_api_version != buildtime_api_version)
    CCTK_VERROR("library version %s, build-time API version 0x%x, run-time API "
                "version 0x%x",
                library_version, buildtime_api_version, runtime_api_version);
  CCTK_VINFO("library version %s, API version 0x%x", library_version,
             buildtime_api_version);
  return 0;
}
```

I don't recall the exact output, but the error path triggers with an API version conflict: Built against 0x020100, found 0x020000 at run time.

I would be happy to provide more information.

**Comment by hzhou** [comment]
Thanks @eschnett . It seems we have a problem building hwloc embedded. What should be done is setting `CFLAGS` with `-fvisibility=hidden` and configure hwloc with `--enable-embedded-mode --disable-visibility`. This, of course, is orthogonal to the fix to use external hwloc.

**Comment by raffenet** [comment]
> Thanks @eschnett . It seems we have a problem building hwloc embedded. What should be done is setting `CFLAGS` with `-fvisibility=hidden` and configure hwloc with `--enable-embedded-mode --disable-visibility`. This, of course, is orthogonal to the fix to use external hwloc.

Visibility has given us trouble on macOS because of the lack of support for weak symbols, so it is always disabled on that platform.

File: configure.ac (hunk 2 above)
Comment (raffenet) [reply configure.ac]: Dropped this part.

**Comment by raffenet** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by raffenet** [comment]
test:jenkins/custom --with-hwloc=/nfs/gce/projects/pmrs/opt/hwloc
netmod: ch3:tcp

**Comment by raffenet** [comment]
test:jenkins/custom --with-hwloc-prefix=/nfs/gce/projects/pmrs/opt/hwloc
netmod: ch3:tcp
